File: pagi-chat-desktop-backend/src/protocol.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

// --- 1. Request from Frontend (User Input) ---
#[derive(Debug, Deserialize, Clone)]
pub struct ChatRequest {
    pub session_id: Uuid,
    pub user_id: String,
    pub timestamp: DateTime<Utc>,
    pub message: String,
}

// --- 2. Response to Frontend (Agent Output) ---
#[derive(Debug, Serialize, Clone)]
#[serde(tag = "type")]
pub enum ChatResponse {
    // For final, complete responses
    #[serde(rename = "complete_message")]
    CompleteMessage {
        id: Uuid,
        content: String,
        is_final: bool,
        latency_ms: u64,
        source_memories: Vec<String>, // RAG sources cited
        issued_command: Option<AgentCommand>,
    },

    // For streaming responses (if desired later)
    #[serde(rename = "message_chunk")]
    MessageChunk {
        id: Uuid,
        content_chunk: String,
        is_final: bool,
    },

    // For status updates (e.g., LLM call failed, memory loaded)
    #[serde(rename = "status_update")]
    StatusUpdate {
        status: String,
        details: Option<String>,
    },
}

// --- 3. Structured Command (Agent controlling the UI) ---
#[derive(Debug, Serialize, Clone)]
#[serde(tag = "command")]
pub enum AgentCommand {
    #[serde(rename = "show_memory_page")]
    ShowMemoryPage { memory_id: Uuid, query: String },

    #[serde(rename = "prompt_for_config")]
    PromptForConfig { config_key: String, prompt: String },

    #[serde(rename = "execute_tool")]
    ExecuteTool {
        tool_name: String,
        arguments: serde_json::Value,
    },
}

File: frontend-digital-twin/types/protocol.ts
// PAGI Chat Protocol Types
// Matches pagi-chat-desktop-backend/src/protocol.rs exactly

// --- 3. Structured Command (Agent controlling the UI) ---
export type AgentCommand =
  | {
      command: 'show_memory_page';
      memory_id: string; // UUID as string
      query: string;
    }
  | {
      command: 'prompt_for_config';
      config_key: string;
      prompt: string;
    }
  | {
      command: 'execute_tool';
      tool_name: string;
      arguments: any; // serde_json::Value equivalent
    };

// --- 1. Request from Frontend (User Input) ---
export interface ChatRequest {
  session_id: string; // UUID as string
  user_id: string; // Matches path param /ws/chat/:user_id
  timestamp: string; // ISO 8601 datetime string (DateTime<Utc>)
  message: string;
  /**
   * True when the operator is actively recording and/or screensharing.
   * Used by the Orchestrator to reason about multi-modal context.
   */
  media_active?: boolean;
}

// --- 2. Response to Frontend (Agent Output) ---
// Discriminated union matching Rust backend's ChatResponse enum

export interface CompleteMessage {
  type: 'complete_message';
  id: string; // UUID as string
  content: string;
  is_final: boolean;
  latency_ms: number;
  source_memories: string[]; // RAG sources cited
  issued_command: AgentCommand | null;
  raw_orchestrator_decision?: string | null;
}

export interface MessageChunk {
  type: 'message_chunk';
  id: string; // UUID as string
  content_chunk: string;
  is_final: boolean;
}

export interface StatusUpdate {
  type: 'status_update';
  status: string; // e.g., "error", "ready", "busy"
  details: string | null;
}

export type ChatResponse = CompleteMessage | MessageChunk | StatusUpdate;
